pulsar-io/kafka/src/main/java/org/apache/pulsar/io/kafka/KafkaBytesSource.java

```diff
@@ -31,85 +31,189 @@
+        if (produceKeyValue) {
+            Object key = extractSimpleValue(consumerRecord.key());
+            Object value = extractSimpleValue(consumerRecord.value());
+            Schema currentKeySchema = getSchemaFromObject(consumerRecord.key(), keySchema);
+            Schema currentValueSchema = getSchemaFromObject(consumerRecord.value(), valueSchema);
+            return new KeyValueKafkaRecord(consumerRecord,
+                    new KeyValue<>(key, value),
+                    currentKeySchema,
+                    currentValueSchema);
+
+        } else {
+            Object value = consumerRecord.value();
+            return new KafkaRecord(consumerRecord,
+                    extractSimpleValue(value),
+                    getSchemaFromObject(value, valueSchema));
+
+        }
+    }
+
+    private static ByteBuffer extractSimpleValue(Object value) {
+        // we have substituted the original Deserializer with
+        // ByteBufferDeserializer in order to save memory copies
+        // so here we can have only a ByteBuffer or at most a
+        // BytesWithKafkaSchema in case of ExtractKafkaAvroSchemaDeserializer
+        if (value == null) {
             return null;
+        } else if (value instanceof BytesWithKafkaSchema) {
+            return ((BytesWithKafkaSchema) value).getValue();
+        } else if (value instanceof ByteBuffer) {
+            return (ByteBuffer) value;
         } else {
-            throw new UnsupportedOperationException("Cannot extract a value from a " + value.getClass());
+            throw new IllegalArgumentException("Unexpected type from Kafka: "+value.getClass());
         }
     }
 
-    @Override
-    public org.apache.pulsar.client.api.Schema<ByteBuffer> extractSchema(ConsumerRecord<Object, Object> consumerRecord) {
-        Object value = consumerRecord.value();
+    private Schema<ByteBuffer> getSchemaFromObject(Object value, Schema fallback) {
         if (value instanceof BytesWithKafkaSchema) {
+            // this is a Struct with schema downloaded by the schema registry
+            // the schema may be different from record to record
             return schemaCache.get(((BytesWithKafkaSchema) value).getSchemaId());
         } else {
-            return Schema.BYTEBUFFER;
+            return fallback;
+        }
+    }
+
+    private static Schema<ByteBuffer> getSchemaFromDeserializerAndAdaptConfiguration(String key, Properties props, boolean isKey) {
+        String kafkaDeserializerClass = props.getProperty(key);
+        Objects.requireNonNull(kafkaDeserializerClass);
+
+        // we want to simply transfer the bytes,
+        // by default we override the Kafka Consumer configuration
+        // to pass the original ByteBuffer
+        props.put(key, ByteBufferDeserializer.class.getCanonicalName());
+
+        Schema<?> result;
+        if (ByteArrayDeserializer.class.getName().equals(kafkaDeserializerClass)
```

Comment:
Use switch case to instead.

Reply:
Unfortunately these are not constants and I cannot use the switch.

I like that we have this strong compile binding. If I write string constants I am losing the reference to the class.

This is not code in hotpath, so using the switch is only a syntax issue.

So you feel strong that we should use the switch?

Reply:
> So you feel strong that we should use the switch?

No, just from the syntax perspective, it works for me.

Reply:
Okay. Thanks for your review.
Can you please help merging this patch?
CI is green